Initial implementation of multi-group cross section generation in Cardinal (#1103)
neams-th-coe/cardinal: doc/content/tutorials/openmc_mgxs.md

# Multi-Group Cross Section Generation

In this tutorial, you will learn how to generate [!ac](MGXS) with distributed cell tallies to couple Cardinal with
deterministic transport codes. To access this tutorial,

```
cd cardinal/tutorials/lwr_mgxs
```

!alert! note title=Previous Experience
In this tutorial, we assume the user is familiar with mesh generation via the
[Reactor module](https://mooseframework.inl.gov/modules/reactor/index.html) and running
[!ac](LWR) calculations with Cardinal-OpenMC. The [LWR AMR tutorial](openmc_amr.md) is
used as a base for this tutorial; we recommend the LWR AMR tutorial for users not familiar with
these concepts.
!alert-end!

## Geometry and Computational Models id=model_1

!include c5g7_ce.md

When generating [!ac](MGXS), tallies do not need to be normalized to power, but Cardinal
still requires a value of reactor power when performing k-eigenvalue calculations. For this tutorial we

> select an arbitrary power of 1 Wth. If you want to compute cross sections accounting for

Reviewer: ```suggestion
select an arbitrary power of 1 W$_\text{th}$. If you want to compute cross sections accounting for
```

multi-physics feedback, a true reactor power and a coupled thermal-hydraulics solver are required.

### OpenMC Model id=openmc

The OpenMC model follows standard model building practices for [!ac](LWR) geometries. The Python script used
to generate the `model.xml` file can be found below, and a plot of the geometry can be found in [assembly_openmc].

!listing /tutorials/lwr_amr/make_openmc_model.py

!media assembly_amr_openmc.png
  id=assembly_openmc
  caption=OpenMC geometry colored by material ID shown on the $x$-$y$ and $x$-$z$ planes
  style=width:90%;margin-left:auto;margin-right:auto

To generate the XML files needed to run OpenMC, you can run the following:

```bash
python make_openmc_model.py
```

Or you can use the `model.xml` file that is included in the `tutorials/lwr_mgxs` directory.

### Mesh Mirror for MGXS Generation id=mesh

Most [!ac](MGXS) workflows store the resulting cross sections in custom libraries for specific deterministic
transport codes. Instead, Cardinal determines homogenization volumes using the mesh
mirror, and stores the cross sections on the mesh at the end of the OpenMC simulation. This process allows for

> "transfering" of [!ac](MGXS) between Cardinal and a coupled deterministic code using the multi-app system.

Reviewer: ```suggestion
"transfer" of [!ac](MGXS) between Cardinal and a coupled deterministic code using the multi-app system.
```

This determininstic code coupling enables higher fidelity cross sections generation and recalculation of group properties to account for multi-physics feedback. The input file used to generate this mesh mirror can be found below.

!listing /tutorials/lwr_mgxs/mesh.i

The mesh is similar to the mesh generated in the [AMR tutorial](openmc_amr.md) with a few key differences to

> accomodate a coupled deterministic transport code. We set `NUM_SECTORS` and `FUEL_RADIAL_DIVISIONS` to `4` to increase

Reviewer: ```suggestion
accommodate a coupled deterministic transport code. We set `NUM_SECTORS` and `FUEL_RADIAL_DIVISIONS` to `4` to increase
```

the radial fidelity of of each pin. The increased discretization is necessary for deterministic calculations as these regions

> have strong gradients due to absorption in the fuel and control rods. We also avoid deleting the gap blocks and

Reviewer: What is the purpose of the "We also avoid deleting the gap blocks" sentence?

extruded the mesh an additional 21.42 cm to cover the reflector. These changes are necessary to preserve mesh connectivity
for deterministic transport solvers using these cross sections, and to ensure we generate cross sections for
the reflector above the core.

!listing /tutorials/lwr_mgxs/mesh.i
  start=[3D_Core]
  end=[Rename]

The input mesh Cardinal will use to store cell tally results on (`mesh_in.e`) can be generated by running the
command below. The resulting mesh can be found in [mgxs_mesh]; we note that this mesh is substantially finer

> in then a conventional Cardinal mesh mirror to adequately capture spatial gradients in the determistic solver.

Reviewer: ```suggestion
than a conventional Cardinal mesh mirror to adequately capture spatial gradients in the deterministic solver.
```

```bash
cardinal-opt -i mesh.i --mesh-only
```

!media mgxs_mesh.png
  id=mgxs_mesh
  caption=The mesh mirror used for computing and storing [!ac](MGXS), shown with an isometric view and sliced on the $x-y$ plane
  style=width:90%;margin-left:auto;margin-right:auto

### Neutronics Input File id=neutronics

The neutronics calculation is performed over the entire assembly by OpenMC and the wrapping of the OpenMC results in MOOSE is performed in
`openmc_mgxs.i`. We begin by defining the mesh (generated in the previous step) used by Cardinal to determine which cell tallies should be constructed for [!ac](MGXS)
generation:

!listing /tutorials/lwr_mgxs/openmc_mgxs.i
  block=Mesh

> Afterwards, we add two auxvariables and auxkernels for post-processing the cross sections -- we use these to compute the scattering ratio (the sum of scattering cross sections divided by the total cross section) in each energy group. The scattering ratio for each group should

Reviewer: ```suggestion
Afterwards, we add two auxvariables and auxkernels for post-processing the cross sections -- we use these to compute the scattering ratio (the sum of scattering cross sections divided by the total cross section) in each energy group. Physically, the scattering ratio for each group should
```

be less than or equal to one; scattering ratios larger than one may result in non-convergence of the deterministic solver or convergence to

> incorrect results. Note that we did not delete the gap blocks in the mesh, but we are block restricting the auxkernels and auxvariables such that

Reviewer: Is there a reason that these blocks in the mesh need to be kept? A restriction in Gnat? If they are not strictly necessary, I think removing them could simplify some of the explanations needed in the tutorial.

Author: It's a restriction in Gnat at the moment. We use Exodus restarts to load the generated cross sections, which necessitates the use of the Cardinal mesh mirror for the deterministic solve (and so all blocks need to be kept to preserve mesh connectivity for the FEM solve).

In the near future I aim to perform the Exodus restarts or Cardinal MGXS generation in a subapp, and use transfers to move the MGXS' over to Gnat. This will enable the use of a different mesh for the deterministic solver. I just haven't gotten around to implementing that capability yet.

they are not computed over the gap region (block 2). This is necessary as the gaps do not receive any tally hits in OpenMC. Therefore, the
cross sections will be zero resulting in a divide by zero when computing scattering ratios.

!listing /tutorials/lwr_mgxs/openmc_mgxs.i
  block=AuxVariables AuxKernels

Next, the [Problem block](Problem/index.md) (with an [OpenMCCellAverageProblem](OpenMCCellAverageProblem.md)) describes the syntax necessary to
replace the normal MOOSE finite element calculation with an OpenMC neutronics solve. We select `cell_level = 1` to ensure that we generate our
cell to element mapping for an OpenMC geometry depth one lower than the root universe. This will result in a distributed cell tally for each unique

> geometric region in the [!ac](LWR) lattice, as opposed to the entire lattice (which is what would happen with `cell_level = 0`). We then specify the number

Reviewer: ```suggestion
geometric region in the [!ac](LWR) lattice, as opposed to the entire lattice (which is what would happen with `cell_level = 0`, because the lattice is filled inside another cell which forms the root universe). We then specify the number
```

> of particles (`particles = 10000`), the number of inactive batches (`inactive_batches = 50`), and the total number of batches (`batches = 1000`).

Reviewer: ```suggestion
of particles, the number of inactive batches, and the total number of batches.
```

Reviewer: Justification: self-evident from the input file what name corresponds to each, and this insulates us if the input file param value ever changes.

!listing /tutorials/lwr_mgxs/openmc_mgxs.i
  start=[Problem]
  end=[Problem/MGXS]

> After specifying the basis problem settings, we move on to setting up the [!ac](MGXS) parameters with the [MGXS block](SetupMGXSAction.md). We specify

Reviewer: ```suggestion
After specifying the basic problem settings, we move on to setting up the [!ac](MGXS) parameters with the [MGXS block](SetupMGXSAction.md). We specify
```

> a distributed cell tally should be used as the [!ac](MGXS) homogenization domain with `tally_type = cell` and the CASMO 2 group structure as the energy group boundaries. We then select `particle = neutron` to set the particle filter to use when computing cross sections, which is useful when OpenMC is performing coupled neutron-photon transport. We then specify `estimator = 'analog'`, which is required when computing

Author: We already do that in the MGXS action (and throw a warning letting the user know we're switching to `analog`). I just thought it would be a good idea to be explicit in the tutorial.
